### DISCUSSION TITLE
React on Github Pages

### DISCUSSION THREAD

**Discussion by electronicjob** [community/community#60881] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

How should I use React.js on Github Pages?

**Comment by ArdeshirV** [comment, marked as answer]
To use React on GitHub Pages, you'll need to follow these steps:

Create a new repository on GitHub: Go to GitHub and create a new repository for your React project.

Set up your local development environment: Make sure you have Node.js and npm (Node Package Manager) installed on your computer. You can download them from the official Node.js website if you don't have them already.

Create a new React app: Open your terminal or command prompt and navigate to the directory where you want to create your React app. Run the following command to create a new React app:

npx create-react-app my-app
This will create a new directory called my-app with a basic React project structure.

Build your React app: Once the app is created, navigate into the project directory by running cd my-app. Then, run the following command to build your React app:

npm run build
This will create an optimized production build of your React app in the build directory.

Deploy to GitHub Pages: Install the gh-pages package, which will help you deploy your React app to GitHub Pages. Run the following command:

npm install gh-pages --save-dev
Configure your package.json file: Open the package.json file in your project directory and add the following properties at the top level:

"homepage": "https://your-username.github.io/your-repo-name",
"scripts": {
  "predeploy": "npm run build",
  "deploy": "gh-pages -d build"
}
Replace your-username with your GitHub username and your-repo-name with the name of your repository.

Deploy your app: In your terminal or command prompt, run the following command to deploy your React app to GitHub Pages:

npm run deploy
This will create a new branch called gh-pages in your repository and push the built files from the build directory to that branch.

Enable GitHub Pages: Go to your repository on GitHub and navigate to the "Settings" tab. Scroll down to the "GitHub Pages" section and select the gh-pages branch as the source for your GitHub Pages site.

Access your deployed app: After a few moments, your React app should be deployed and accessible at the URL mentioned in the "GitHub Pages" section of your repository settings.

That's it! Your React app is now deployed on GitHub Pages. Anytime you make changes to your app, repeat steps 4 and 7 to rebuild and redeploy your app.

Hope you got your answer.

**Reply by electronicjob** [reply]
thank you!

**Reply by ThinkOpenly** [reply]
Which steps are "steps 4 and 7" ?
(I presume "npm run build" and "npm run deploy", but there are no step numbers listed in the post as I see it.)

And, thanks so much for this post! It worked flawlessly for me!

**Reply by DannyMancheno** [reply]
For those who might have gotten stuck on 'npm run deploy, This will create a branch called gh-pages in your ... ' 

You may have come across a 'Failed to get remote.origin.url' error in your terminal. 
You may want to initialize a local git repository inside your react-app directory (git init)  as well as give it a remote origin pointer to the github repository  (git remote add origin <github-repo-cloning-link>) 

Afterwords, 'npm run build' may succeed, if we had the same issue.

**Reply by albertomtferreira** [reply]
Thank you! It works perfectly.

**Reply by manoj0222** [reply]
@DannyMancheno

i  had done git init and git add command but still facing.May be these was not clear answered

**Comment by atmcpnk** [comment]
> To use React on GitHub Pages, you'll need to follow these steps:
> 
> Create a new repository on GitHub: Go to GitHub and create a new repository for your React project.
> 
> Set up your local development environment: Make sure you have Node.js and npm (Node Package Manager) installed on your computer. You can download them from the official Node.js website if you don't have them already.
> 
> Create a new React app: Open your terminal or command prompt and navigate to the directory where you want to create your React app. Run the following command to create a new React app:
> 
> npx create-react-app my-app This will create a new directory called my-app with a basic React project structure.
> 
> Build your React app: Once the app is created, navigate into the project directory by running cd my-app. Then, run the following command to build your React app:
> 
> npm run build This will create an optimized production build of your React app in the build directory.
> 
> Deploy to GitHub Pages: Install the gh-pages package, which will help you deploy your React app to GitHub Pages. Run the following command:
> 
> npm install gh-pages --save-dev Configure your package.json file: Open the package.json file in your project directory and add the following properties at the top level:
> 
> "homepage": "https://your-username.github.io/your-repo-name", "scripts": { "predeploy": "npm run build", "deploy": "gh-pages -d build" } Replace your-username with your GitHub username and your-repo-name with the name of your repository.
> 
> Deploy your app: In your terminal or command prompt, run the following command to deploy your React app to GitHub Pages:
> 
> npm run deploy This will create a new branch called gh-pages in your repository and push the built files from the build directory to that branch.
> 
> Enable GitHub Pages: Go to your repository on GitHub and navigate to the "Settings" tab. Scroll down to the "GitHub Pages" section and select the gh-pages branch as the source for your GitHub Pages site.
> 
> Access your deployed app: After a few moments, your React app should be deployed and accessible at the URL mentioned in the "GitHub Pages" section of your repository settings.
> 
> That's it! Your React app is now deployed on GitHub Pages. Anytime you make changes to your app, repeat steps 4 and 7 to rebuild and redeploy your app.
> 
> Hope you got your answer.

Question, does the gh-pages branch need all of the files to be pushed into it? currently, in my situation I have a 'dist' folder inside of the build folder which is what is displaying inside of my gh-pages branch. On top of this I also have a second 'dist' folder in the root directory. not sure which im supposed to keep and where.

**Reply by BashkirovN** [reply]
No, only the contents of "dist" are required. The "dist" which is created by `npm run build` command and is located at the root of your project.

**Comment by aliansari180** [comment]
I have deploy react website on GitHub pages but my images not showing i have images in public folder.
Please help me

**Reply by BashkirovN** [reply]
In my app, I have images in the `public/img/` folder and the path looks like `src="/img/icon.png"`

**Reply by leeringrose** [reply]
@aliansari180 I think if those images are good to be displayed on your localhost launch, it's associated to deployment stuff.
They are not displayed even on your localhost, it's about your image's path given in the code or image format problem.

**Reply by dragontheory** [reply]
Same issue here! 

Was anyone able to find a working solution? 

Browser console shows 404 for `assets/img`. Images are deployed to `/public/assets/img/`, but the pointers for those images are to `/assets/img/`.

My deployed URL: https://d7460n.github.io/hud_react/
My GitHub repo URL: https://github.com/D7460N/hud_react

This blog post suggests that there is an issue with deploying a React app with routing to GitHub Pages. Solution did not work. 
https://blog.logrocket.com/deploying-react-apps-github-pages/

The author of the template I am using suggests adding the basename to the BrowserRouter in `index.js`. I am deploying to a subdirectory called `hud_react`. 
Example: `<BrowserRouter basename="/hud_react/">`
https://themeforest.net/item/hud-react-18-bootstrap-admin-template/42252040/comments

**Comment by Dhirajchaudari** [comment]
Please suggest me branch structure for GitHub Pages. Should we create one main branch where all our source code is present ? and do we have to run build cmd inside of the "main" branch ? I am asking you this because I need to update the website after hosting so how should I do that 
Please help me

**Reply by ThinkOpenly** [reply]
In my project, I currently have two branches: a "main" branch, and a topic branch. I can deploy either. It depends on the context in which `npm run deploy` is executed.

**Reply by leeringrose** [reply]
In my case, usually use `gh-pages` branch in creating it newly aside with `main` branch and keep that new branch updated for every `main` branch.
